### ISSUE TITLE
Segfault when BiPredicate in javalib, but works when in sanbox/Test

### ISSUE THREAD

**Issue by pbatko** [scala-native/scala-native#1229]
I wanted to port `java.util.function.BiFunction`, but it fails when I put the code is in the javalib project.
On the other hand, when I put it directly in `sandbox/Test.scala`[2] it compiles and runs fine.
@Duhemm 

[1]
```
[success] Total time: 0 s, completed Apr 15, 2018 2:24:05 PM
[info] Updating {file:/home/pbatko/src/scala-native/}sandbox...
[info] Resolving jline#jline;2.14.3 ...
[info] Done updating.
[info] Compiling 1 Scala source to /home/pbatko/src/scala-native/sandbox/target/scala-2.11/classes...
[info] Linking (784 ms)
[info] Discovered 1292 classes and 9504 methods
[info] Optimizing (debug mode) (1138 ms)
[info] Generating intermediate code (301 ms)
[info] Produced 40 files
[info] Compiling to native code (1034 ms)
[info] Linking native code (boehm gc) (129 ms)
java.lang.RuntimeException: Nonzero exit code: 139
	at scala.sys.package$.error(package.scala:27)
	at scala.scalanative.sbtplugin.ScalaNativePluginInternal$$anonfun$scalaNativeConfigSettings$6$$anonfun$apply$8$$anonfun$apply$9$$anonfun$apply$10.apply(ScalaNativePluginInternal.scala:153)
	at scala.scalanative.sbtplugin.ScalaNativePluginInternal$$anonfun$scalaNativeConfigSettings$6$$anonfun$apply$8$$anonfun$apply$9$$anonfun$apply$10.apply(ScalaNativePluginInternal.scala:153)
	at scala.Option.foreach(Option.scala:236)
	at scala.scalanative.sbtplugin.ScalaNativePluginInternal$$anonfun$scalaNativeConfigSettings$6$$anonfun$apply$8$$anonfun$apply$9.apply(ScalaNativePluginInternal.scala:153)
	at scala.scalanative.sbtplugin.ScalaNativePluginInternal$$anonfun$scalaNativeConfigSettings$6$$anonfun$apply$8$$anonfun$apply$9.apply(ScalaNativePluginInternal.scala:138)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
[trace] Stack trace suppressed: run last sandbox/compile:run for the full output.
[error] (sandbox/compile:run) Nonzero exit code: 139
[error] Total time: 6 s, completed Apr 15, 2018 2:24:12 PM
```

[2]
```
import java.util.function.Function

import java.util.function.BiFunction
// // NOTE move this to javalib project to reproduce
//trait BiFunction[T, U, R] {
//  self =>
//
//  def andThen[V](other: java.util.function.Function[R, V]): BiFunction[T, U, V] = {
//    if (other == null) throw new NullPointerException
//
//    new BiFunction[T, U, V] {
//      override def apply(t: T, u: U): V = {
//        val r = self.apply(t, u)
//        other.apply(r)
//
//      }
//    }
//  }
//
//  def apply(t: T, u: U): R
//
//}


object Test {
  def main(args: Array[String]): Unit = {

    val bifunction = new BiFunction[Int, String, List[String]] {
      override def apply(t: Int, u: String): List[String] = {
        List(t.toString, u)
      }
    }

    val function = new java.util.function.Function[List[String], String] {
      override def apply(t: List[String]): String = t.mkString("[", ",", "]")
    }

    val result = bifunction.andThen[String](function).apply(1, "abc")
    println(result)

  }

}
```

**Comment by densh** [comment]
I originally assumed that this one is another instance of #375, but it turns out it's a bit more fun than that. 

When you compile the `Test` object, scalac sees `BiFunction` as a java-defined interface so it doesn't apply the same treatment as it would for a Scala trait by using the trait's underlying implementation class that contains all default method implementations. So it means that currently you can not provide implementations for default methods in JDK.

So in the end, it's blocked by #233.

**Comment by ekrich** [comment]
@sjrd I am running into this on the JUnit PR #1932 - Formerly it was commented out - since this has been removed from the `0.4.0` milestone, do I need to comment out the test again? I am assuming there is no good `2.11` resolution.

The `java.util.function.BiFunctionTest` will not link.

**Comment by sjrd** [comment]
The only solution for 2.11 would be to implement the hacky mechanism of `@JavaDefaultMethod` that Scala.js has...

**Comment by ekrich** [comment]
Aren't we hoping that we will drop 2.11 soon so we could just leave that test commented out as it was before?

**Comment by sjrd** [comment]
That's what we need to do at the moment, yes.

I'm not sure about dropping 2.11 soon. We'll need to see how the ecosystem of libraries reacts first.

**Comment by ekrich** [comment]
I meant soon as when we get Scala 3 support - hopefully we will have people on 2.12 or greater by then. 😄

**Comment by WojciechMazur** [comment]
Fixed in #2042
